src/main/java/com/Elessar/app/Main.java

```diff
@@ -1,49 +1,101 @@
+
+    /**
+     *
+     * @param args [0] username for user1
+     *             [1] port number for user1
+     *             [2] username for user2
+     *             [3] port number for user2
+     *             [4] total number of messages to send
+     */
+    public static void main(String[] args) throws Exception {
+        final String serverURL = new StringBuilder().append("http://127.0.0.1:9000").toString();
+        final MetricManager metricManager = new MetricManager("ClientMetric", 0);
+
+        final String username1 = args[0];
+        final String password1 = "qwe123052vawqw";
+        final String phone1 = getRandomNum(10);
+        final BlockingQueue<String> messageQueue1 = new LinkedBlockingQueue<>();
+        final int port1 = Integer.valueOf(args[1]);
+
+        final String username2 = args[2];
+        final String password2 = "mnsdg21g1742fz";
+        final String phone2 = getRandomNum(10);
+        final BlockingQueue<String> messageQueue2 = new LinkedBlockingQueue<>();
+        final int port2 = Integer.valueOf(args[3]);
+
+        int messageCount = Integer.valueOf(args[4]);
+
+        final MyClient client1 = new MyClient(serverURL, metricManager);
+        final MyClientServer clientServer1 = new MyClientServer("localhost", port1, messageQueue1, metricManager);
+        final MyClient client2 = new MyClient(serverURL, metricManager);
+        final MyClientServer clientServer2 = new MyClientServer("localhost", port2, messageQueue2, metricManager);
+
+        clientServer1.run();
+        clientServer2.run();
+
+        client1.register(username1, password1, username1 + "@163.com", phone1);
+        client2.register(username2, password2, username2 + "@163.com", phone2);
+
+        client1.logOn(username1, password1, port1);
+        client2.logOn(username2, password2, port2);
+
+        final Random r = new Random();
 
-    private static void testProtoBuf() {
-        final Person.Builder sendPersonData = Person.newBuilder().setId(123).setName("DIBL").setEmail("abc@gmail.com");
-        final Person.PhoneNumber.Builder phoneNumber = Person.PhoneNumber.newBuilder().setNumber("321-7684231");
-        sendPersonData.addPhones(phoneNumber);
-
-        try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
-            sendPersonData.build().writeTo(os);
-            try (InputStream is = new ByteArrayInputStream(os.toByteArray())) {
-                final Person receivePersonData = Person.parseFrom(is);
-                System.out.printf("receive and send data is equal: %b\n", sendPersonData.equals(receivePersonData));
-                System.out.println(receivePersonData.toString());
-                System.out.println(sendPersonData.toString());
+        while (messageCount > 0) {
+            int msgCount1 = r.nextInt(5) + 1;
+            while (messageCount > 0 && msgCount1 > 0) {
+                client1.sendMessage(username1, username2, getRandomStr(10));
```

Comment:
Maybe we want to do some timing here as well, like avg / p90 time of how long does a client request get a response.
Also we want to keep track how many requests failed. If failure rate is high then result of the perf test is not reliable

Reply:
> Maybe we want to do some timing here as well, like avg / p90 time of how long does a client request get a response.
Why not use metric measured from MyClient for this performance?

Reply:
You are right... I just forgot we had metrics in MyClient. 
Ok we don't need metrics in perfMain, just keep track of error is enough. My bad.